### PR TITLE
Backend Bug and Build Fixes for RotatedSPOs when QMC_COMPLEX=1

### PR DESCRIPTION
**Pull request by rcclay**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes
Currently, RotatedSPOs has guards to make sure it is executed only in the real build.  Attempts to run calculations when removing these guards caused build errors due to type inconsistencies in the data structure definitions and in the math kernels.  I have fixed these issues and removed the guards, which allows a lot of the math kernels and RotatedSPOs API's to be tested (and pass) in the QMC_COMPLEX build.  This is still guarded at the SPOSet builder level, so users can't hit it, but at least it will be tested.  

More specifically:
1.) Removed QMC_COMPLEX guard for RotatedSPOs, so it will always be built and executed.  
2.) Promoted all "param" types to ValueType, whereas before there was a mixture of ValueType and RealType.  
3.) Generalized exponentiate_antisym_matrix and log_antisym_matrix calls to a.) execute the correct expressions for complex, and b.) exhibit correct behavior for real and complex matrices.  
4.) Did the same for constructing the antisymmetric/antihermitian matrices.  
5.) Added additional tests for log and exp matrix functions to test the complex generalizations.  

There were also some "bug" fixes.  exponentiate_antisym_matrix had a row-major/column major issues which came out in the wash for the real build, but not for complex matrices.  

What I did NOT do:
1.) Parameter handling between real and complex.  Unit tests that loaded parameters from the h5 file had dimensionality issues, likely due to the 2x difference in numbers between real and complex.  There's a guard for this now.
2.) Fix multideterminant API's.  There's some baked in real vs. complex type issues at a deep level...  fixing that will be a future PR.
3.) Check derivatives and values of complex wave functions.  This will be done in a future PR advertising a working complex orb opt capability.  For now, I only care that the real cases pass in the complex build.  

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix
- Refactoring (no functional changes, no api changes)
- Build related changes
- Testing changes (e.g. new unit/integration/performance tests)
- Documentation changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Intel Xeon, RHEL8
## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by rcclay** [comment]
Going to regroup with GCC.  Was using intel oneAPI, and I wonder if it let some stuff slide.

File: src/QMCWaveFunctions/RotatedSPOs.cpp
Comment (camelto2) [inline src/QMCWaveFunctions/RotatedSPOs.cpp:290]: this should just be deleted, not commented out

File: src/QMCWaveFunctions/RotatedSPOs.cpp
Comment (camelto2) [inline src/QMCWaveFunctions/RotatedSPOs.cpp:315]: delete, not comment

File: src/QMCWaveFunctions/RotatedSPOs.cpp
Comment (camelto2) [inline src/QMCWaveFunctions/RotatedSPOs.cpp:320]: delete, not comment out

File: src/QMCWaveFunctions/RotatedSPOs.cpp
Comment (camelto2) [inline src/QMCWaveFunctions/RotatedSPOs.cpp:399]: delete, not comment

File: src/QMCWaveFunctions/RotatedSPOs.cpp
Comment (camelto2) [inline src/QMCWaveFunctions/RotatedSPOs.cpp:519]: I need to think through everything going on here. It looks like there is a relative transpose happening between what you're doing here vs. previously. Is this now consistent with what we've been working out in our notes?

File: src/QMCWaveFunctions/RotatedSPOs.h
Comment (camelto2) [inline src/QMCWaveFunctions/RotatedSPOs.h:123]: Just leaving a note, based on our previous discussion this will likely need to end up being changed back to RealType since it interfaces with SPOSetBuilder, and SPOSetBuilder explicitly passes this in as RealType. However, this can be fixed when we fully enable complex

File: src/QMCWaveFunctions/RotatedSPOs.cpp
Comment (camelto2) [reply src/QMCWaveFunctions/RotatedSPOs.cpp:519]: I assume by keeping it as mat[j][i] in the arguments, it produces the wrong matrix exponentials and logs in the tests?

Comment (rcclay) [reply src/QMCWaveFunctions/RotatedSPOs.cpp:519]: Actually, lets block this for now.  As it is written now, it passes the unit tests which were compared against scipy.linalg, whereas before it didn't (matrix was conjugated when it shouldn't have been).  However, lets make sure this doesn't break the real code finite difference tests of the derivative vs. parameter variations.

Comment (rcclay) [reply src/QMCWaveFunctions/RotatedSPOs.cpp:515]: Just checked it.  All good.

File: src/QMCWaveFunctions/RotatedSPOs.cpp
Comment (rcclay) [reply src/QMCWaveFunctions/RotatedSPOs.cpp]: Done

File: src/QMCWaveFunctions/RotatedSPOs.cpp
Comment (rcclay) [reply src/QMCWaveFunctions/RotatedSPOs.cpp]: Done

File: src/QMCWaveFunctions/RotatedSPOs.h
Comment (rcclay) [reply src/QMCWaveFunctions/RotatedSPOs.h:123]: Agreed.  For now, this is coded to conform to current complex handling of optimizable types.  When we decide to do something about it, I will fix this and the parameter dependent unit tests.

File: src/QMCWaveFunctions/tests/test_RotatedSPOs.cpp
Comment (markdewing) [inline src/QMCWaveFunctions/tests/test_RotatedSPOs.cpp:403]: It would be nice if this example (and the log) were added to `gen_matrix_ops.py`.  However, that script needs some changes to make it easier to add new examples, so maybe not for this PR.

Comment (rcclay) [reply src/QMCWaveFunctions/tests/test_RotatedSPOs.cpp:403]: I actually hacked gen_matrix_ops.py to generate these reference values, but it was kinda nasty and ad hoc and I was ashamed to add it.

Comment (markdewing) [reply src/QMCWaveFunctions/tests/test_RotatedSPOs.cpp:403]: That happens.  With the reorganized code in #4818, making a cleaner version should be easier.    Creating a new function `create3x3_complex_matrix` that uses numpy (instead of Sympy) should work.  (And some of output code may need updating, depending on whether you want to just print the values vs.  outputting them in a format ready to use in the unit tests)

Comment (rcclay) [reply src/QMCWaveFunctions/tests/test_RotatedSPOs.cpp:403]: Oh that's a LOT better.  I can either add it after #4818 is merged prior to this merge, or I can do a follow up PR after they are both in.

Comment (markdewing) [reply src/QMCWaveFunctions/tests/test_RotatedSPOs.cpp:403]: If there's no urgency to get this PR in, wait for #4818, and then update this PR.

Comment (rcclay) [reply src/QMCWaveFunctions/tests/test_RotatedSPOs.cpp:403]: Added reference value generation to gen_matrix_ops.py.

File: src/QMCWaveFunctions/CMakeLists.txt
Comment (markdewing) [inline src/QMCWaveFunctions/CMakeLists.txt:43]: Commented-out code should be removed.

**Comment by rcclay** [comment]
After looking into the H5 parameter packing/unpacking, found a type issue I had missed.  This allows us to eliminate the QMC_COMPLEX guards on the parameter handling, so essentially the entire unit test passes now.

**Comment by rcclay** [comment]
Let's make this contingent on upcoming PR to make optimizable parameters real.

File: src/QMCWaveFunctions/tests/CMakeLists.txt
Comment (camelto2) [inline src/QMCWaveFunctions/tests/CMakeLists.txt:157]: delete

**Comment by rcclay** [comment]
OK!  @jptowns fixed the complex VariableSet -> real VariableSet issues.  If this passes the tests, it should be good to go.

**Comment by ye-luo** [comment]
Note that this fix build RotatedSPO in the complex code. At the current stage only real part of complex parameters are optimized.

File: src/QMCWaveFunctions/RotatedSPOs.cpp
Comment (ye-luo) [inline src/QMCWaveFunctions/RotatedSPOs.cpp:634]: Could you reduce duplication?

Comment (rcclay) [reply src/QMCWaveFunctions/RotatedSPOs.cpp:634]: This function was a pain...  in contrast to exponentiate_antisym_matrix and its LAPACK::heev call, this function relies on geev, which has a different argument list if its real or complex.  I could make one function if I could use #ifdef QMC_COMPLEX or if constexpr.

File: src/QMCWaveFunctions/RotatedSPOs.cpp
Comment (PDoakORNL) [inline src/QMCWaveFunctions/RotatedSPOs.cpp:684]: Since RealType is real guaranteed here why not just use the literals?

Still it is less awful than
`RealType cone(1.0);`
Which combines this idiom usually only used with types that can be real or complex with a variable name that I once assumed meant `complex 1` in QMCPACK

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
In additional to only optimizing the real part, the other issue here is that you can't read in for the parameters to start with.  I kind of feel like this could have should have waited for the complex/real templating but we have some communication problems.  That said other than need to document what I think is basically a temporary hack in `setRotationalParameters` I don't see any reason this can't be merged if its a useful intermediate step.

File: src/QMCWaveFunctions/RotatedSPOs.cpp
Comment (rcclay) [reply src/QMCWaveFunctions/RotatedSPOs.cpp]: Done

**Comment by rcclay** [comment]
@PDoakORNL This is meant to be an intermediate step towards supporting full complex (complex optimizable parameters) orbital optimization.  Trying to split these up into more readable chunks, but even this one is kinda beefy.

**Review by PDoakORNL — DISMISSED** [review]

File: src/QMCWaveFunctions/RotatedSPOs.cpp
Comment (camelto2) [inline src/QMCWaveFunctions/RotatedSPOs.cpp:38]: Just leaving a note that this will need to be converted as a proper unpacking into ValueType in the coming PRs.

File: src/QMCWaveFunctions/RotatedSPOs.cpp
Comment (camelto2) [inline src/QMCWaveFunctions/RotatedSPOs.cpp:39]: Just leaving a note that this will need to be converted to a proper unpacking into ValueType in a subsequent PR. Beyond the scope of what is being done here though

**Comment by prckent** [comment]
Test this please

**Comment by prckent** [comment]
(May show CI failures due to problems with the CI on ornl-sulfur V100; we are debugging)

**Comment by prckent** [comment]
Test this please

**Review by ye-luo — CHANGES_REQUESTED** [review]
Please reduce the code duplication as discussed in
https://github.com/QMCPACK/qmcpack/pull/4816/files#r1395759349

**Comment by rcclay** [comment]
@ye-luo I have one function for log_antisym_matrix, with complex vs. type divergences handled by #ifdef qmc_complex directives.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — DISMISSED** [review]
LGTM.

**Comment by prckent** [comment]
Trouble with complex offload in CI

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
> Trouble with complex offload in CI

Caused by feature not yet ready. https://github.com/QMCPACK/qmcpack/issues/4843

**Pull request merged**
